### PR TITLE
fix: improve outside interactions (click outside)

### PR DESCRIPTION
**Pull request by huntabyte**
Closes: #865 #750 #832

This is the initial introduction to improving interactions within Melt. This closes a few different outstanding bugs, so I think we should review/release this bit here to take care of those.

In the near future, I'd like to work with you, @TGlide to see how we can better encapsulate the "modal" behavior across the various builders that use it. We should be able to incorporate all of the things that modals handle:
- closing on outside interactions/clicks
- preventing scroll
- handling escape keydown events
- focus traps
- etc...

Essentially, the goal is to make this a single action that can handle all the above cases, with their respective overrides, handler functions, conditions, etc. so we can fix something in one place and it becomes a fix across all builders. We can then expose simple/consistent APIs to builder consumers without having to put too much thought into it, and behavior is familiar/consistent across every builder.

### CODE REVIEW

File: src/lib/builders/dialog/create.ts [hunk 1]
```diff
@@ -228,17 +206,16 @@ export function createDialog(props?: CreateDialogProps) {
 				}),
 
 				effect([closeOnOutsideClick, open], ([$closeOnOutsideClick, $open]) => {
-					return useClickOutside(node, {
-						enabled: $open,
-						handler: (e: PointerEvent) => {
+					return useModal(node, {
```
Comment (TGlide) [inline src/lib/builders/dialog/create.ts:209]: Could we maybe have it receive the stores directly? That way we don't need to do an effect here.

File: src/lib/internal/actions/modal/action.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+		const index = visibleModals.indexOf(node);
+		if (index >= 0) {
+			visibleModals.splice(index, 1);
+		}
+	}
+
+	function update(config: ModalConfig) {
+		const { open, onClose, shouldCloseOnInteractOutside, closeOnInteractOutside } = config;
+
+		sleep(100).then(() => {
```
Comment (TGlide) [inline src/lib/internal/actions/modal/action.ts:21]: Doesn't tick or sleep(1) suffice here?

Comment (huntabyte) [reply src/lib/internal/actions/modal/action.ts:21]: Nice catch, let me take another look to ensure the tests still pass without this. I know it was added during debugging the nested dialogs, so I may have let it sneak in by accident!

Comment (huntabyte) [reply src/lib/internal/actions/modal/action.ts:21]: Actually, I left it as it was originally in your implementation of handling the nested dialogs:

CleanShot 2024-01-11 at 10 27 02@2x

We can see if the tests still pass with it being 1 or less than 100 though!

File: src/lib/builders/dialog/create.ts (hunk 1 above)
Comment (huntabyte) [reply src/lib/builders/dialog/create.ts:209]: So you're thinking we pass in the stores and just move the `effect`s into the `useModal`? I think that's a good idea and would clean this up quite a bit.

File: src/lib/internal/actions/modal/action.ts (hunk 2 above)
Comment (huntabyte) [reply src/lib/internal/actions/modal/action.ts]: Okay I removed this and it passed all the tests 🫡

File: src/lib/builders/dialog/create.ts (hunk 1 above)
Comment (huntabyte) [reply src/lib/builders/dialog/create.ts]: Consolidated the effect into the action

File: src/lib/internal/actions/interact-outside/action.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+import type { InteractOutsideConfig, InteractOutsideEvent } from './types.js';
+
+export function useInteractOutside(node: HTMLElement, config: InteractOutsideConfig) {
+	let unsub = noop;
+	function update(config: InteractOutsideConfig) {
+		unsub();
+		const { onInteractOutside, onInteractOutsideStart, enabled } = config;
+
+		if (!enabled) {
+			return {
```
Comment (TGlide) [inline src/lib/internal/actions/interact-outside/action.ts:11]: Why are you returning destroy here? You should make unsub = noop, and then do an early return, no?

File: src/lib/internal/actions/interact-outside/types.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+export type InteractOutsideEvent = PointerEvent | MouseEvent | TouchEvent;
+
+export type InteractOutsideConfig = {
+	onInteractOutside?: (e: InteractOutsideEvent) => void;
```
Comment (TGlide) [inline src/lib/internal/actions/interact-outside/types.ts:4]: If the other one is start, what is this one? End? Maybe some JSDoc here would help clear things up

**Comment by babakfp** [comment]
It seems like #832 and #865 are not fixed:

https://github.com/melt-ui/melt-ui/assets/44144724/9dc9ff52-b334-4b7c-99bc-11474ffc3980

I tested it in https://huntabyte-fix-outside-intera.melt-ui.pages.dev/docs/builders/dialog

I'm not sure if this PR is sposed to fix the issue in the above URL or not.

**Comment by huntabyte** [comment]
Putting this back into draft while I figure out where I introduced the regression because it should be preventing the behavior being displayed above.

**Comment by huntabyte** [comment]
@TGlide I'm having some issues with moving the effect into the action itself, seems to have something to do with cleanup as the events aren't unregistering, so I reverted it back to the effect within the `create` dialog.

@babakfp, the browser's mobile emulation won't cut it here as we're detecting the possibility of `PointerEvent`s, which your PC's browser still has. If you clone the repo serve it via ngrok and access it via your phone, you'll see that it does in fact close those issues.

**Comment by babakfp** [comment]
> @babakfp, the browser's mobile emulation won't cut it here as we're detecting the possibility of `PointerEvent`s, which your PC's browser still has. If you clone the repo serve it via ngrok and access it via your phone, you'll see that it does in fact close those issues.

Testing it with my phone:

https://github.com/melt-ui/melt-ui/assets/44144724/21006703-380a-43cc-895c-688949230cdb

**Comment by huntabyte** [comment]
Same for this? https://a71888c7.melt-ui.pages.dev/

**Comment by babakfp** [comment]
> Same for this? https://a71888c7.melt-ui.pages.dev/

Yep:


https://github.com/melt-ui/melt-ui/assets/44144724/b68ab5b1-76f5-420c-9843-850aca1f95ee

**Comment by huntabyte** [comment]
Browser and OS? @babakfp

**Comment by babakfp** [comment]
> Browser and OS? @babakfp

Chrome. Android.

IMG_20240119_054226
IMG_20240119_054205
IMG_20240119_054141

**Comment by huntabyte** [comment]
Latest preview I can't reproduce

https://github.com/melt-ui/melt-ui/assets/64506580/94880b72-f8b1-4807-93f8-9904ecb32715

**Pull request merged**